returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,226 @@
+    data processing work across multiple CPU cores using `MultiProcDataset` and in
+    turn frees the GPU from data preprocessing tasks.
+
+    Both functions can be specified at the same time, and their results will be
+    combined by first applying the segment-level function and then the
+    multiple-segments-level function.
+
+    Example usage::
+        from returnn.tensor.dim import Dim, DimTypes
+
+        time_dim = Dim(None, kind=DimTypes.Spatial)
+        new_data_dim = Dim(128)
+
+        train = {
+            "class": "PostprocessingDataset",
+            "dataset": {
+                "class": "HDFDataset",
+                "files": ["/path/to/data.hdf"],
+            },
+            # one of them, but not both:
+            "map_seq": map_seq,  # (data: TensorDict) -> TensorDict
+            "map_seq_stream": map_seqs,  # (iter: Iterator[TensorDict]) -> Iterator[TensorDict]
+            # only required when data shapes change wrt. the wrapped dataset:
+            "map_outputs": {
+                "data": {"dims": [time_dim, new_data_dim]},
+            },
+        }
+    """
+
+    def __init__(
+        self,
+        dataset: Dict[str, Any],
+        map_seq: Optional[Union[Callable[[TensorDict], TensorDict]]] = None,
+        map_seq_stream: Optional[Callable[[Iterator[TensorDict]], Iterator[TensorDict]]] = None,
+        map_outputs: Optional[Dict[str, Any]] = None,
+        _meta_info_cache: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+        """
+        :param dataset: inner dataset to be post-processed
+        :param map_seq: post processor function operating on the single-segment level.
+        :param map_seq_stream: post processor function operating on the multiple segment level via an iterator.
+            Allows merging multiple segments into one, or generating multiple output segments from one input segment.
+            When both functions are specified, the segment-level function is applied first to every segment, and the
+            results are passed to the iterator-level function.
+        :param map_outputs: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+        :param _meta_info_cache: internal usage
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        if self.seq_ordering != "default":
+            raise ValueError(f"specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}")
+        if map_seq is None and map_seq_stream is None:
+            raise ValueError("need to either set map_seq or map_seq_stream")
+        if map_seq and map_seq_stream:
+            raise ValueError("cannot set both map_seq and map_seq_stream")
```

Comment:
```suggestion
        if self.seq_ordering != "default":
            raise ValueError(f"{self}: specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}")
        if map_seq is None and map_seq_stream is None:
            raise ValueError(f"{self}: need to either set map_seq or map_seq_stream")
        if map_seq and map_seq_stream:
            raise ValueError(f"{self}: cannot set both map_seq and map_seq_stream")
```
Btw, when I make code suggestions here, usually you could (or even should) just overtake them directly (there is a button here for that, just click on it), unless you have good reasons to do it differently.